Random::ISAAC: fix initial seed

asterite opened the pull request:
Fixes #7976

In the old code `unsafe_as` returned a copy of the static array and so it was initializing something else other than what was returned.

asterite commented on spec/std/random/isaac_spec.cr:
```diff
@@ -349,4 +349,10 @@ describe "Random::ISAAC" do
+  it "different instances generate different numbers (#7976)" do
+    isaacs = Array.new(1000) { Random::ISAAC.new }
+    values = isaacs.map(&.rand(10_000_000))
+    values.uniq.size.should be > 2
```
This spec might fail, but it's veeeeery unlikely.

asterite replied:
Any other way to test this? Otherwise I can remove the test altogether.

asterite commented on src/random/isaac.cr:
```diff
@@ -12,13 +12,10 @@ class Random::ISAAC
-    result_slice = result.unsafe_as(StaticArray(UInt8, sizeof(Seeds))).to_slice
-    Random::Secure.random_bytes(result_slice)
-    result
+    StaticArray(UInt32, 8).new do
```
This is probably a bit slower than before (assuming the old code did it right) but this is on initialization and I prefer safe code even if a bit slower (not in hot path).

Blacksmoke16 commented on src/random/isaac.cr:
```diff
@@ -12,13 +12,10 @@ class Random::ISAAC
-    Random::Secure.random_bytes(result_slice)
-    result
+    StaticArray(UInt32, 8).new do
+      Random::Secure.rand(0_u32..UInt32::MAX)
```
```suggestion
      Random::Secure.rand(UInt32::MIN..UInt32::MAX)
```

ysbaddaden commented:
What about an uninitialized `UInt8[32]`, fill it in one call to `random_bytes` instead of 8 calls to rand, then cast it to a `StaticArray(UInt32, 8)`?

That would spare 7 unbuffered reads from urandom (or 7 calls to getrandom, arc4random, ...) yet fix the issue, wouldn't it?

asterite commented:
I think so, yes, but to do that we need to use "unsafe_as". I wanted to avoid unsafe code. But I'll try it anyway.

Another idea is to have such code in Secure::Random, somehow, so that it's unsafe but in a single place.

asterite commented:
Updated!

I added a general `Random#rand(StaticArray(T, N))` that, for now, only works with integer types. Its implementation uses `unsafe_as` but it's not really unsafe (or it shouldn't be).

Then I used that method in the initialization of `Random::ISSAC` seed.

In the future I'd also like to be able to do `rand(Int32)`, `rand(UInt64)` and so on.

asterite commented:
I'll do it differently.

asterite commented:
Added more things and implemented it in a way that's easier to read in docs (now there's one overload per static array type).

Also added the ability to do `rand(Int32)`, `rand(Int8)` and so on, to get a value in that type's range. For that I use `random_bytes` to fill all bytes and then cast it to that type (same thing as with static arrays). I don't know if that's sound but we could eventually change the implementation if needed.

Sija commented on src/random.cr:
```diff
@@ -289,26 +289,39 @@ module Random
+                           "UInt32".id => %w(3147957137 4245108745 2207809043 3184391838),
+                           "Int64".id  => %w(4438449217673515190 8514493061600538358 -4874671083204037318 -7825896160729246667),
+                           "UInt64".id => %w(15004487597684511003 12027825265648206103 11303949506191212698 6228566501671148658),
+                         } %}
```
Since these numbers will change anyway isn't it pointless to include them in docs?

asterite replied:
I wanted to include example values. How would you do it otherwise?

Sija replied:
I'd write the values like `<random {{type}} integer>`.

asterite replied:
I think a real example is better

Sija commented on src/random.cr:
````diff
@@ -289,6 +289,41 @@ module Random
+    # ```
+    # rand({{type}}) # => {{values[0].id}}
+    # ```
+    def rand(t : {{type}}.class) : {{type}}
````
```suggestion
    def rand(type : {{type}}.class) : {{type}}
```

asterite replied:
Thanks, fixed

oprypin commented on src/random.cr:
````diff
@@ -289,6 +289,41 @@ module Random
+    # rand({{type}}) # => {{values[0].id}}
+    # ```
+    def rand(type : {{type}}.class) : {{type}}
+      rand_type_from_bytes(type)
````
It's strange to have a separate implementation for this when `rand(type::MIN..type::MAX)` works fine.

Also kind of funny that `random_bytes` unrolls integers into bytes (and inverts for endinanness) and then this rolls them back into an integer (without inverting for endianness - though double-inverting would also be weird).

So yes, this implementation is endianness-dependent, unlike everything else.

And then, just by chance,
`Random.new(x).rand(Int32::MIN..Int32::MAX) == Random.new(x).rand(Int32)`
but
`Random.new(x).rand(Int64::MIN..Int64::MAX) != Random.new(x).rand(Int64)`
